webapp/tests/karma/ts/services/training-cards.service.spec.ts

```diff
@@ -0,0 +1,460 @@
+import { TestBed } from '@angular/core/testing';
+import { provideMockStore } from '@ngrx/store/testing';
+import sinon from 'sinon';
+import { expect } from 'chai';
+
+import { GlobalActions } from '@mm-actions/global';
+import { DbService } from '@mm-services/db.service';
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { ModalService } from '@mm-modals/mm-modal/mm-modal';
+import { TrainingCardsService } from '@mm-services/training-cards.service';
+import { SessionService } from '@mm-services/session.service';
+import { RouteSnapshotService } from '@mm-services/route-snapshot.service';
+
+describe('TrainingCardsService', () => {
+  let service: TrainingCardsService;
+  let globalActions;
+  let xmlFormsService;
+  let dbService;
+  let modalService;
+  let localDb;
+  let clock;
+  let consoleErrorMock;
+  let sessionService;
+  let routeSnapshotService;
+
+  beforeEach(() => {
+    localDb = { allDocs: sinon.stub() };
+    dbService = { get: () => localDb };
+    globalActions = { setTrainingCard: sinon.stub(GlobalActions.prototype, 'setTrainingCard') };
+    xmlFormsService = { subscribe: sinon.stub() };
+    modalService = { show: sinon.stub() };
+    sessionService = {
+      isDbAdmin: sinon.stub(),
+      userCtx: sinon.stub(),
+    };
+    consoleErrorMock = sinon.stub(console, 'error');
+    routeSnapshotService = { get: sinon.stub() };
+
+    TestBed.configureTestingModule({
+      providers: [
+        provideMockStore(),
+        { provide: DbService, useValue: dbService },
+        { provide: XmlFormsService, useValue: xmlFormsService },
+        { provide: ModalService, useValue: modalService },
+        { provide: SessionService, useValue: sessionService },
+        { provide: RouteSnapshotService, useValue: routeSnapshotService },
+      ]
+    });
+
+    service = TestBed.inject(TrainingCardsService);
+  });
+
+  afterEach(() => {
+    clock && clock.restore();
+    sinon.restore();
+    document
+      .querySelectorAll('#enketo-test')
+      .forEach(element => element.remove());
+  });
+
+  it('should show uncompleted training form when none are completed', async () => {
+    sessionService.userCtx.returns({ roles: [ 'chw' ], name: 'a_user' });
+    localDb.allDocs.resolves({ rows: [] });
+    clock = sinon.useFakeTimers(1653312565642); // 23/05/2022 20:29:25
+    const xforms = [
+      {
+        _id: 'abc-123',
+        internalId: 'training:form-a',
```

Comment:
To make the test more comprehensive, please shuffle the trainings, so the "soonest" one it not also the first that gets returned, and we should we intentionally select the training card with the lowest `start_date`.

Reply:
Done